src/pojo-interfaces.ts

```diff
@@ -0,0 +1,422 @@
+/* eslint-disable license-header/header */
+/**
+ * This is just a strawman proposal;
+ * the actual new interfaces will not be implemented through this file.
+ *
+ * They're also not set in stone. Actual implementation experience might result in further tweaks.
+ * Names can still be changed if we propose it as an actual specification, of course.
+ *
+ * The design goals of these interfaces are:
+ * - JSON.parse(JSON.stringify(data)) is lossless
+ * - Optimised for reading and copying in an expected shape in stateful (Solid) applications
+ * - fromRdfJsDataset(toRdfJsDataset(data)) is lossless
+ *
+ * Non-goals:
+ * - Optimising for streaming data
+ * - Optimising for unknown use cases
+ * - Be perfect from the get-go
+ * - Be immediately useful outside of solid-client
+ * @module
+ */
+
+import { dataset as rdfJsDataset } from "@rdfjs/dataset";
+import RdfJsDataFactory from "@rdfjs/data-model";
+import * as RdfJs from "rdf-js";
+import { IriString } from "./interfaces";
+import { XmlSchemaTypeIri, xmlSchemaTypes } from "./datatypes";
+
+const localNodeSkolemPrefix = "https://inrupt.com/.well-known/sdk-local-node/" as const;
+type LocalNodeIri = `${typeof localNodeSkolemPrefix}${string}`;
+export type LocalNodeName = string;
+type DataTypeIriString = XmlSchemaTypeIri | IriString;
+type LocaleString = string;
+export type Objects = Readonly<
+  Partial<{
+    literals: Readonly<Record<DataTypeIriString, readonly string[]>>;
+    langStrings: Readonly<Record<LocaleString, readonly string[]>>;
+    namedNodes: ReadonlyArray<LocalNodeIri | IriString>;
+    // By abstracting over the specific blank nodes, we can neatly skip over the
+    // issue of fetching the same data twice and then not being able to reconcile
+    // different instances of the same blank nodes.
+    blankNodes: readonly Predicates[];
+    // TODO: Do we need RDF/JS Variables as well, or are they just for SPARQL?
+  }>
+>;
+
+export type Predicates = Readonly<Record<IriString, Objects>>;
+
+export type Subject = Readonly<{
+  type: "Subject";
+  url: LocalNodeIri | IriString;
+  predicates: Predicates;
+}>;
+
+export type Graph = Readonly<Record<IriString, Subject>>;
+
+export type Dataset = Readonly<{
+  type: "Dataset";
```

Comment:
So as discussed today, this seems like a straightforward suggestion (i.e., to reflect the overall intent of the PR):
```suggestion
export type ImmutableDataset = Readonly<{
  type: "ImmutableDataset";
```

Reply:
6fc0730754b8c3efd3168a868298697e4052e1fd